modules/twinklefluff.js

```diff
@@ -80,101 +80,97 @@ Twinkle.fluff = {
 	},
 
 	diff: function() {
-		if( mw.config.get('wgCanonicalSpecialPageName') === "Undelete" ) {
-			//You can't rollback deleted pages!
-			return;
-		}
+		if (mw.config.get('wgDiffNewId') === mw.config.get('wgRevisionId')) {
+			// Add a [restore this revision] link to the older revision
+			// Don't show if there's a single revision or weird diff (cur on latest)
+			if (mw.config.get('wgDiffOldId') && (mw.config.get('wgDiffOldId') !== mw.config.get('wgDiffNewId'))) {
+				var revertToRevision = document.createElement('div');
+				revertToRevision.setAttribute( 'id', 'tw-revert-to-orevision' );
+				revertToRevision.style.fontWeight = 'bold';
+
+				var revertToRevisionLink = buildLink('SaddleBrown', 'restore this version');
+				revertToRevisionLink.href = "#";
+				$(revertToRevisionLink).click(function(){
+					Twinkle.fluff.revertToRevision(mw.config.get('wgDiffOldId').toString());
+				});
+				revertToRevision.appendChild(revertToRevisionLink);
 
-		var otitle, ntitle;
-		try {
-			var otitle1 = document.getElementById('mw-diff-otitle1');
-			var ntitle1 = document.getElementById('mw-diff-ntitle1');
-			if (!otitle1 || !ntitle1) {
-				return;
+				var otitle = document.getElementById('mw-diff-otitle1').parentNode;
+				otitle.insertBefore( revertToRevision, otitle.firstChild );
+			}
+
+			// Add either restore or rollback links to the newer revision
+			// Don't show if there's a single revision or weird diff (prev on first)
+			var ntitle = document.getElementById('mw-diff-ntitle1').parentNode;
+			if( document.getElementById('differences-nextlink') ) {
+				// Not latest revision
+				var revertToRevisionN = document.createElement('div');
+				revertToRevisionN.setAttribute( 'id', 'tw-revert-to-nrevision' );
+				revertToRevisionN.style.fontWeight = 'bold';
+
+				var revertToRevisionNLink = buildLink('SaddleBrown', 'restore this version');
+				revertToRevisionNLink.href = "#";
+				$(revertToRevisionNLink).click(function(){
+					Twinkle.fluff.revertToRevision(mw.config.get('wgDiffNewId').toString());
+				});
+				revertToRevisionN.appendChild(revertToRevisionNLink);
+
+				ntitle.insertBefore( revertToRevisionN, ntitle.firstChild );
+			} else if( Twinkle.getPref('showRollbackLinks').indexOf('diff') !== -1 && mw.config.get('wgDiffOldId') && (mw.config.get('wgDiffOldId') !== mw.config.get('wgDiffNewId') || document.getElementById('differences-prevlink'))) {
```

Comment:
This `else if` is somewhat unwieldy.  I can make it shorter, but I think the code is likely easier to maintain if we handle the left-hand restore first then tackle the right-hand separately, adding either restore or rollback links.

I'd like to try and deduplicate even more of this down the road.